Support IPv4 parsing with leading zeros on go 1.17

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Since go version 1.17, net.ParseIP() and net.ParseCIDR() functions rejects leading zeros in the dot-decimal notation of IPv4 addresses.

The affected fields in the VMI spec are:
- dnsConfig.nameservers (k8s.io.api.core.v1.PodDNSConfig)
- domain.devices.interfaces[].dhcpOptions.ntpServers (v1.DHCPOptions)
- networks[].pod.vmNetworkCIDR

These fields will be parsed as decimal IPv4 addresses (leading zeros are ignored):
For example: 192.168.010.1 => 192.168.10.1
Same for CIDR notation:
For example: 192.168.010.1/24 => 192.168.10.1/24

In order to upgrade Kubevirt's go version to 1.17 and to maintain backwards compatibility - we need to use ParseIPSloppy() and ParseCIDRSloppy() functions from k8s.io/utils.
Because we don't want to have a mix of the go standard library functions and the sloppy functions from K8s, the chosen solution uses them only **once** in the unmarshaling methods of the three affected structs.
The UnmarshalJSON methods make sure that the affected fields are sanitized before their usage in the system.

This was tested manually and automatically on go 1.17 (locally).

Several other ideas where considered:
1. Change the validating webhook and the three affected places where string IP/CIDR is converted to bytes - this can lead to a mix to standard and sloppy functions that we want to avoid. Furthermore, it is not simple to determine when to use which function, and unforeseen bugs could occur.
2. Change all occurrences of net.ParseIP() and net.ParseCIDR() to ParseIPSloppy() and ParseCIDRSloppy().
This can lead to more problems when interfacing with other non Go tools or systems (that may have other interpretation of IPv4 addresses with leading zeros).
3. Using a mutating webhook - this option is not feasible because the webhook is not being called after a system upgrade from Kubevirt that had been built using go < 1.17 to a version build using go >= 1.17.

For additional details please see issue #6498.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6498

**Special notes for your reviewer**:
~~This PR is based on PR #7008, please review the last five commits.~~

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

orelmisan commented:
/cc @EdDev @ormergi

orelmisan commented:
/uncc @enp0s3 @vladikr

orelmisan commented:
Tested the fix on current go version (1.16.6) with the following VMI manifest:

```yaml
---
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-fedora
  name: vmi-fedora
spec:
  dnsConfig:
    nameservers:
      - 01.1.1.1
  domain:
    devices:
      interfaces:
        - masquerade: {}
          name: default
          dhcpOptions:
            ntpServers:
              - 162.159.200.0123
              - 010.35.255.14
      disks:
      - disk:
          bus: virtio
        name: containerdisk
      - disk:
          bus: virtio
        name: cloudinitdisk
      rng: {}
    resources:
      requests:
        memory: 1024M
  networks:
    - name: default
      pod: {}
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:
      image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
    name: containerdisk
  - cloudInitNoCloud:
      userData: |-
        #cloud-config
        password: fedora
        chpasswd: { expire: False }
    name: cloudinitdisk
```

Run the following command inside the VM:
```bash
[fedora@vmi-fedora ~]$ journalctl | grep -Ei 'dhcp'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.7147] dhcp-init: Using DHCP client 'internal'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8161] dhcp4 (eth0): activation: beginning transaction (timeout in 45 seconds)
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8245] dhcp4 (eth0): option dhcp_lease_time      => '86313600'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8248] dhcp4 (eth0): option domain_name          => 'default.svc.cluster.local'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8248] dhcp4 (eth0): option domain_name_servers  => '10.96.0.10 1.1.1.1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8249] dhcp4 (eth0): option domain_search        => 'default.svc.cluster.local svc.cluster.local cluster.local'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8249] dhcp4 (eth0): option expiry               => '1726481865'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8250] dhcp4 (eth0): option host_name            => 'vmi-fedora'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8250] dhcp4 (eth0): option interface_mtu        => '1480'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8251] dhcp4 (eth0): option ip_address           => '10.0.2.2'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8251] dhcp4 (eth0): option ntp_servers          => '162.159.200.123 10.35.255.14'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8251] dhcp4 (eth0): option requested_broadcast_address => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8252] dhcp4 (eth0): option requested_domain_name => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8252] dhcp4 (eth0): option requested_domain_name_servers => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8253] dhcp4 (eth0): option requested_domain_search => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8253] dhcp4 (eth0): option requested_host_name  => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8254] dhcp4 (eth0): option requested_interface_mtu => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8254] dhcp4 (eth0): option requested_ms_classless_static_routes => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8254] dhcp4 (eth0): option requested_nis_domain => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8255] dhcp4 (eth0): option requested_nis_servers => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8255] dhcp4 (eth0): option requested_ntp_servers => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8256] dhcp4 (eth0): option requested_rfc3442_classless_static_routes => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8256] dhcp4 (eth0): option requested_root_path  => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8257] dhcp4 (eth0): option requested_routers    => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8257] dhcp4 (eth0): option requested_static_routes => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8258] dhcp4 (eth0): option requested_subnet_mask => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8258] dhcp4 (eth0): option requested_time_offset => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8259] dhcp4 (eth0): option requested_wpad       => '1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8259] dhcp4 (eth0): option routers              => '10.0.2.1'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8260] dhcp4 (eth0): option subnet_mask          => '255.255.255.0'
Dec 22 10:17:45 vmi-fedora NetworkManager[640]: <info>  [1640168265.8260] dhcp4 (eth0): state changed unknown -> bound
```

orelmisan commented:
Inside the virt-launcher pod:
```bash
$ cat /etc/resolv.conf
search default.svc.cluster.local svc.cluster.local cluster.local
nameserver 10.96.0.10
nameserver 01.1.1.1
```

orelmisan commented:
/retest-required

orelmisan commented:
/retest-required

ormergi reviewed (commented):
Hi, I have few general comments about the PR,
- I think you can merged commits 4-6 to the third one, they all about the same things its just touching different files, also its better to keep unit tests code on the same commit of the actual change.
- Could you please add to the PR description brief explanation about how this PR changes affect the behavior (e.g: before this change on VM create ... from now on it will go like..) 

Regarding second commit, what you think about wrapping IP address parser calls?
I think that adding another abstraction layer in this case will make the code easier to understand (by adding a comment about what this function does and why).

About the last commit, what you think about adding another test for leading zero's input addresses input instead of changing the one we have?
Also I would add another test the example on the issue this PR refer to (e.g: verify that `192.168.010.1` interprets to `192.168.10.1` instead  `192.168.8.1`)

EdDev commented on pkg/network/dhcp/server/server.go:
```diff
@@ -156,7 +158,7 @@ func prepareDHCPOptions(
 
 			for _, server := range customDHCPOptions.NTPServers {
-				ip := net.ParseIP(server).To4()
+				ip := netutils.ParseIPSloppy(server).To4()
```
I think that changing this here may cause a maintenance problem in the future.
It will be much better to normalize/adjust the values as soon as possible in the setup flow. Then, if any changes will get introduced later on (refactoring or others), it will still work as expected.

If I am not mistaken, this code is in the setup flow, so we should able to cover it early on (e.g. at `netconf`).
Can you please explore this option?

orelmisan replied:
Sure.

EdDev commented on tests/network/vmi_networking.go:
```diff
@@ -483,7 +483,7 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
 				BootFileName:   "config",
 				TFTPServerName: "tftp.kubevirt.io",
-				NTPServers:     []string{"127.0.0.1", "127.0.0.2"},
+				NTPServers:     []string{"162.159.200.0123", "010.35.255.14"},
```
Why did you change the original loopback addresses?
Using real addresses may cause traffic to exit the test setup and result in unexpected behavior. We not always know what is the test setup (e.g. kubevirtci is only one of them).

orelmisan replied:
Reverted to original IP addresses.

EdDev reviewed (changes requested):
The first part, about the validators seems good to me.

The second part has some issues:
- Unit tests should come in the same commit as the production code changed.
- The usage of the sloppy-parser deep inside the setup flow is a bit problematic IMO. See inline comment about the ability to maintain such a thing. It will be better to normalize these values early on the network setup flow.
- It seems we are missing e2e tests for the namservers IP/s.

orelmisan commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:
```diff
@@ -33,6 +33,8 @@ import (
 	"k8s.io/apimachinery/pkg/util/validation"
 	k8sfield "k8s.io/apimachinery/pkg/util/validation/field"
 
+	netutils "k8s.io/utils/net"
```
Thanks, fixed  it.

orelmisan commented:
Changes:
1. Unified production code and unit tests to same commit.
2. Added E2E test for DNS nameservers.

orelmisan commented:
/retest-required

orelmisan commented:
/retest-required

EdDev commented on pkg/network/dns/resolveconf_test.go:
```diff
@@ -18,6 +18,14 @@ var _ = Describe("Resolveconf", func() {
 			Expect(err).To(BeNil())
 		})
 
+		It("should allow parsing of IPv4 nameservers with leading zeros", func() {
+			ns1, ns2 := []uint8{8, 8, 8, 8}, []uint8{8, 8, 4, 4}
+			resolvConf := "nameserver 08.8.8.08\nnameserver 8.08.04.4\n"
+			nameservers, err := ParseNameservers(resolvConf)
+			Expect(nameservers).To(Equal([][]uint8{ns1, ns2}))
+			Expect(err).To(BeNil())
+		})
```
I am aware that you have just copy/paste it from the previous test, but:
- Please separate between expectations and the input data.
- There is no need to check the error independently, this will also work:
`Expect(ParseNameservers(resolvConf)).To(Equal([][]uint8{ns1, ns2}))`

orelmisan replied:
This is not copy/paste, I've used an existing test and added leading zeros to the input fields.
I prefer to make this change in another PR.

orelmisan replied:
Done.

EdDev commented on tests/network/vmi_networking.go:
```diff
@@ -520,7 +520,7 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
 			dnsVMI.Spec.DNSPolicy = "None"
 			dnsVMI.Spec.DNSConfig = &k8sv1.PodDNSConfig{
-				Nameservers: []string{"8.8.8.8", "4.2.2.1"},
+				Nameservers: []string{"8.8.8.8", "04.2.2.01"},
```
`04` and `01` are not valid?
The parser is forbidding leading zeros or translating them as octal and only when the range is false it fails?

orelmisan replied:
Every octet is parsed as decimal:
04 is parsed as 4, 0127 is parsed as 127.
I've added an example to the PR description as @ormergi had suggested.

EdDev replied:
So this will not fail with the new parsing behavior, because both 1 and 4 are equivalent in value to the octet version.

orelmisan replied:
That is right.

EdDev reviewed (commented):
Regarding the commit messages: Please arrange their first line to fit this format: <topic>: <subject>

The topic is usually the package under which this change occured (e.g. `network, dhcp`, `network, dns`, `webhooks`, etc)

orelmisan commented:
> Regarding the commit messages: Please arrange their first line to fit this format: `<topic>: <subject>`
> 
> The topic is usually the package under which this change occured (e.g. `network, dhcp`, `network, dns`, `webhooks`, etc)

@EdDev  Done.

orelmisan commented:
/test pull-kubevirt-unit-test

orelmisan commented:
> About the last commit, what you think about adding another test for leading zero's input addresses input instead of changing the one we have?

@ormergi Done.

orelmisan commented:
Change: Added dedicated E2E tests for DNS nameservers and NTP IPv4 addresses which contain leading zeros.

orelmisan commented:
This PR will be split and rebased on PR #7008

orelmisan commented:
Changes:
1. Rebased on PR #7008.
2. Added spacial UnmarshalJSON methods for VirtualMachineSpec and DHCPOptions structs.

This was tested manually and automatically on go 1.17 (locally).

orelmisan commented:
Change:
Changed the signature of SanitizeIP() to return an error as well as a sanitized value.
This was done in order not to assign "<nil>" strings to DNS / NTP array elements - so the user can have an indication of what went wrong.

EdDev commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -156,6 +156,28 @@ type VirtualMachineInstanceSpec struct {
+		return err
+	}
+
+	*vmiSpec = VirtualMachineInstanceSpec(vmiSpecAlias)
```
I think you should first mutate the alias and only then assign it at the very end.

orelmisan replied:
Done.

EdDev reviewed (commented):
Nice, thanks!

The sanitize commit introduces a function that no one uses there, therefore you better squash it to the commit after, where it is used.

EdDev commented:
Please add a convincing argument why it is the appropriate to do this at the unmarshal step.

orelmisan commented on staging/src/kubevirt.io/api/core/v1/sanitizers.go:
```diff
@@ -0,0 +1,32 @@
+/*
+Copyright 2021 The KubeVirt Authors.
```
Changed to the proper license.

orelmisan commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -156,6 +156,28 @@ type VirtualMachineInstanceSpec struct {
+	var vmiSpecAlias VMISpecAlias
+
+	err := json.Unmarshal(data, &vmiSpecAlias)
+	if err != nil {
```
Done.

orelmisan commented:
Change:
1. Fixed code review remarks.
2. Added handling for networks[].pod.vmNetworkCIDR field.

orelmisan commented:
/retest-required

orelmisan commented:
Change: Fixed a bug in the custom CIDR E2E test.

orelmisan commented:
/retest-required

EdDev commented on staging/src/kubevirt.io/api/core/v1/sanitizers.go:
```diff
@@ -33,3 +33,13 @@ func SanitizeIP(address string) (string, error) {
+func SanitizeCIDR(cidr string) (string, error) {
+	ip, net, err := netutils.ParseCIDRSloppy(cidr)
+	if err != nil {
+		return "", fmt.Errorf("not a valid CIDR")
```
What about the original error? What does it contain that requires you to add your own text?

orelmisan replied:
You are right, there is no need to use a custom error.

EdDev commented on staging/src/kubevirt.io/api/core/v1/sanitizers.go:
```diff
@@ -0,0 +1,35 @@
+	netutils "k8s.io/utils/net"
+)
+
+func SanitizeIP(address string) (string, error) {
```
Why is this public?

orelmisan replied:
Changed both sanitizeIP and SanitizeCIDR to be un-exported.

orelmisan commented on staging/src/kubevirt.io/api/core/v1/sanitizers.go:
```diff
@@ -0,0 +1,35 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
```
Done.

orelmisan commented:
Change: Fixed review remarks.

EdDev reviewed (approved):
Thank you, looks great.

It is now ready to negotiate with the maintainer if we should take this change directly (+1 from my side).
Note: If we do take it, the two commits that are canceling each other can be taken out.

EdDev commented:
/sig network

orelmisan commented:
/retest-required

orelmisan commented:
rebase (to pass pull-kubevirt-generate test)
@EdDev can you please lgtm again?

orelmisan commented:
/retest-required

AlonaKaplan commented:
> 3. Using a mutating webhook - this option is not feasible because the webhook is not being called during migration.

Why do you need to mutate the IP during a migration?

AlonaKaplan commented on tests/network/vmi_networking.go:
```diff
@@ -677,20 +682,23 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
 				}
 
 				By("Checking ping (IPv4) to gateway")
-				ipAddr := gatewayIPFromCIDR(networkCIDR)
```
Now the custom networkCIDR is ignored.

orelmisan replied:
Because the current implementation of the tests doesn't support getting the GW address from a CIDR with leading zeros on go 1.17 - It fails at gatewayIPFromCIDR().
It could be solved by using ParseCIDRSloppy(), but as @EdDev had said, we don't want to legitimize usage of this function.

orelmisan replied:
ParseIPSloppy() and ParseCIDRSloppy() are used only once in the code (sanitizers.go)

AlonaKaplan replied:
I find avoiding the `ParseCIDR` with this workaround is very confusing and hiding the real reason.
We don't really need an extra parameter for a custom gateway IP address. It is just a trick to avoid the `ParseCIDR`.
I prefer to be straight forward and use `ParseCIDRSloppy()`.

orelmisan replied:
Done.

orelmisan commented:
> > 3. Using a mutating webhook - this option is not feasible because the webhook is not being called during migration.
> 
> Why do you need to mutate the IP during a migration?

We want to handle the scenario of a system upgrade, on which a migration happens.

AlonaKaplan commented:
> > > 3. Using a mutating webhook - this option is not feasible because the webhook is not being called during migration.
> > 
> > 
> > Why do you need to mutate the IP during a migration?
> 
> We want to handle the scenario of a system upgrade, on which a migration happens.

I meant that is case of migration, the VM already has the mutated version of the IP. So i don't understand the reasoning of "mutating webhook is not being called during migration"?

orelmisan commented:
> > > > 3. Using a mutating webhook - this option is not feasible because the webhook is not being called during migration.
> > > 
> > > 
> > > Why do you need to mutate the IP during a migration?
> > 
> > 
> > We want to handle the scenario of a system upgrade, on which a migration happens.
> 
> I meant that is case of migration, the VM already has the mutated version of the IP. So i don't understand the reasoning of "mutating webhook is not being called during migration"?

The problem here is the upgrade from a Kubevirt version that had been built using go < 1.17 to a version that had been built using go >= 1.17.

During this transition, you can still have IPv4 addresses and CIDRs with leading zeros stored in the DB.

AlonaKaplan commented:
> > > > > 3. Using a mutating webhook - this option is not feasible because the webhook is not being called during migration.
> > > > 
> > > > 
> > > > Why do you need to mutate the IP during a migration?
> > > 
> > > 
> > > We want to handle the scenario of a system upgrade, on which a migration happens.
> > 
> > 
> > I meant that is case of migration, the VM already has the mutated version of the IP. So i don't understand the reasoning of "mutating webhook is not being called during migration"?
> 
> The problem here is the upgrade from a Kubevirt version that had been built using go < 1.17 to a version that had been built using go >= 1.17.
> 
> After this transition, you can still have IPv4 addresses and CIDRs with leading zeros stored in the DB.
> 
> EDIT: So a mutating webhook won't be called on the VMI.

Please update the PR message. "Migration" is not related to the issue.

orelmisan commented:
[Changes:](https://github.com/kubevirt/kubevirt/compare/42e130ee2f30bc0aa633e77584bf99a0f6039626..22ef82d088e55239079f3047a0cf576609b09afd)
1. Removed two unused validating webhook commits.
2. Fixed a remark from @AlonaKaplan's review.

AlonaKaplan commented:
/approve
Thanks!

orelmisan commented:
/resetst-required

orelmisan commented:
/retest-required

orelmisan commented:
/retest-required

orelmisan commented:
/test pull-kubevirt-e2e-k8s-1.21-sig-storage

EdDev commented:
/test pull-kubevirt-e2e-kind-1.22-sriov

rmohr commented:
/cherrypick release-0.49

I think it is the safest to prepare our long lived branch for go 1.17. I am pretty sure that adopters will sooner or later just start using 1.17> and may not know about this issue here.
